source/common/common/utility.h

```diff
@@ -632,4 +633,111 @@ template <class Value> struct TrieLookupTable {
+//   - the starting address of data will not be 8-byte aligned. This is not required
+//     by the C++ standard for a uint8_t, but may be suboptimal on some processes.
+//   - the 6 bytes of data will be at byte offsets 10 to 15, and bytes 16 to 23 will be
+//     unused. This may be surprising to some users, and suboptimal in resource usage.
+// One possible tweak is to declare data_ as a uint64_t[], or to use an `alignas`
+// declaration. As always, micro-optimizations should be informed by
+// microbenchmarks, showing the benefit.
+class InlineStorage : public NonCopyable {
+public:
+  // Custom delete operator to keep C++14 from using the global operator delete(void*, size_t),
+  // which would result in the compiler error:
+  // "exception cleanup for this placement new selects non-placement operator delete"
+  static void operator delete(void* address) { ::operator delete(address); }
+
+protected:
+  /**
+   * @param object_size the size of the base object; supplied automatically by the compiler.
+   * @param data_size the amount of variable-size storage to be added, in bytes.
+   * @return a variable-size object based on data_size_bytes.
+   */
+  static void* operator new(size_t object_size, size_t data_size_bytes) {
+    return ::operator new(object_size + data_size_bytes);
+  }
+};
+
+class InlineString;
+using InlineStringPtr = std::unique_ptr<InlineString>;
+
+// Represents immutable string data, keeping the storage inline with the
+// object. These cannot be copied or held by value; they must be created
+// as unique pointers.
+//
+// Note: this is not yet proven better (smaller or faster) than std::string for
+// all applications, but memory-size improvements have been measured for one
+// application (Stats::SymbolTableImpl). This is presented here to serve as an
+// example of how to use InlineStorage.
+class InlineString : public InlineStorage {
+public:
+  /**
+   * @param str the string_view for which to create an InlineString
+   * @return a unique_ptr to the InlineString containing the bytes of str.
+   */
+  static InlineStringPtr create(absl::string_view str) {
+    return InlineStringPtr(new (str.size()) InlineString(str.data(), str.size()));
+  }
+
+  /**
+   * @return a std::string copy of the InlineString.
+   */
+  std::string toString() const { return std::string(data_, size_); }
+
+  /**
+   * @return a string_view into the InlineString.
+   */
+  absl::string_view toStringView() const { return absl::string_view(data_, size_); }
+
+  /**
+   * @return the number of bytes in the string
+   */
+  size_t size() const { return size_; }
+
+  /**
+   * @return a pointer to the first byte of the string.
+   */
+  const char* data() const { return data_; }
+
+private:
+  // Constructor is declared private so that no one constructs one without the
+  // proper size allocation. to accommodate the variable-size buffer.
+  InlineString(const char* str, size_t size) : size_(size) { memcpy(data_, str, size); }
+
+  size_t size_;
```

Comment:
If we are really going nutty here for size, can't this be uint32_t and save 4 bytes on 64-bit machines?